### PR TITLE
Added /station route

### LINKED ISSUE #128: Confirm API design for read operations
These read operations are essential for building any kind of front end, and I think we should have four main endpoints for now.

# Base URL

`dallasbymetro.com/api/`

# Endpoints

## **GET:** `public/all`

Fetches all data points including Stations and their Points of Interest, grouped by Station.

**Request Parameters**

 - None

**Example Request**

`dallasbymetro.com/api/public/all`

<details>
<summary>Example Response</summary>

```json
{
  "stations": [
    {
      "stationId": 1,
      "name": "Mockingbird Station",
      "servedLines": ["Orange Line", "Red Line"],
      "latitude": 32.8481,
      "longitude": -96.7871,
      "pointsOfInterest": [
        {
          "poiId": 1,
          "name": "Meadows Museum",
          "latitude": 32.8481,
          "longitude": -96.7871,
          "walkingDistance": 7,
          "pic_url": "https://images.meadowsmuseum.com/museum.jpg",
          "type": "Museum",
          "amenities": [
            "Vegetarian Options",
            "Bike Parking",
            "Vegan Options"
          ]
        },
        {
          "poiId": 2,
          "name": "Greenville Avenue Pizza Company",
          "latitude": 32.8480,
          "longitude": -96.7872,
          "walkingDistance": 10,
          "pic_url": "https://images.greenvillepizzaco.com/restaurant.jpg",
          "type": "Restaurant",
          "amenities": [
            "Vegetarian Options",
            "Bike Parking",
            "Alcohol Options"
          ]
        }
      ]
    },
    {
      "stationId": 2,
      "name": "Cityplace/Uptown Station",
      "servedLines": ["Blue Line"],
      "latitude": 32.8482,
      "longitude": -96.7872,
      "pointsOfInterest": [
        {
          "poiId": 3,
          "name": "Uptown Theater",
          "latitude": 32.8482,
          "longitude": -96.7873,
          "walkingDistance": 6,
          "pic_url": "https://images.uptowntheater.com/theater.jpg",
          "type": "Theater",
          "amenities": [
            "Vegetarian Options",
            "Bike Parking",
            "Alcohol Options"
          ]
        },
        {
          "poiId": 4,
          "name": "Beyond Sushi",
          "latitude": 32.8484,
          "longitude": -96.7874,
          "walkingDistance": 10,
          "pic_url": "https://images.beyondsushi.com/restaurant.jpg",
          "type": "Restaurant",
          "amenities": [
            "Vegetarian Options",
            "Bike Parking"
          ]
        }
      ]
    }
  ]
}
```
</details>

## **GET:** `public/stations`

Fetches Station information, including served lines, rail travel time, Points of Interest (POIs), and the POIs' types, walking distances, images, and amenities. The result can be filtered by acceptable rail and walk times, and required POI amenities.

**Request Parameters**
 - current_station (optional): This is required in conjunction with rail_time. This is the starting station that will be used to calculate what stations are reachable from the current_station in the rail_time provided. 
 - rail_time (optional): The acceptable amount of time spent on rail travel (format as string, e.g., "30m" for 30 minutes, "2h" for 2 hours)
 - walk_time (optional): The acceptable amount of time spent walking (format as string, e.g., "10m" for 10 minutes, "1h" for 1 hour)
 - amenities (optional): The required amenities that need to be present at the POI (as a comma-separated list of strings)

**Example Request**

`dallasbymetro.com/api/public/stations?CurrentStation=1&RailTime=30m&WalkTime=10m&Amenities=Vegetarian%Options,Bike%Parking`

<details>
<summary>Example Response</summary>

```json
{
"stations": [
    {
      "stationId": 1,
      "name": "Mockingbird Station",
      "servedLines": ["Orange Line", "Red Line"],
      "pointsOfInterest": [
        {
          "poiId": 1,
          "name": "Meadows Museum",
          "type": "Museum",
          "walkingDistance": "7m",
          "pic_url": "https://images.meadowsmuseum.com/museum.jpg",
          "amenities": ["Vegetarian Options", "Bike Parking", "Vegan Options"]
        },
        {
          "poiId": 2,
          "name": "Greenville Avenue Pizza Company",
          "type": "Restaurant",
          "walkingDistance": "10m",
          "pic_url": "https://images.greenvillepizzaco.com/restaurant.jpg",
          "amenities": ["Vegetarian Options", "Bike Parking", "Alcohol Options"]
        }
      ]
    },
    {
      "stationId": 2,
      "name": "Cityplace/Uptown Station",
      "servedLines": ["Blue Line"],
      "railTravelTime": "30m",
      "pointsOfInterest": [
        {
          "poiId": 3,
          "name": "Uptown Theater",
          "type": "Theater",
          "walkingDistance": "6m",
          "pic_url": "https://images.uptowntheater.com/theater.jpg",
          "amenities": ["Vegetarian Options", "Bike Parking", "Alcohol Options"]
        },
        {
          "poiId": 4,
          "name": "Beyond Sushi",
          "type": "Restaurant",
          "walkingDistance": "10m",
          "pic_url": "https://images.beyondsushi.com/restaurant.jpg",
          "amenities": ["Vegetarian Options", "Bike Parking"]
        }
      ]
    }
  ]
}
```
</details>

## **GET:** `public/pois`

Fetches detailed information about Points of Interest (POIs) based on their ID(s).

**Request Parameters**

 - ID (required): The ID(s) of the Point(s) of Interest you would like all the information for (as a comma-separated list of integers)

**Example Request**

`dallasbymetro.com/api/public/pois?ID=1,2,3`

<details>
<summary>Example Response</summary>

```json
{
  "pointsOfInterest": [
    {
      "poiId": 1,
      "name": "Meadows Museum",
      "type": "Museum",
      "latitude": 32.8481,
      "longitude": -96.7871,
      "walkingDistance": "7m",
      "pic_url": "https://images.meadowsmuseum.com/museum.jpg",
      "amenities": ["Vegetarian Options", "Bike Parking", "Vegan Options"],
      "stationId": 1
    },
    {
      "poiId": 2,
      "name": "Greenville Avenue Pizza Company",
      "type": "Restaurant",
      "latitude": 32.8480,
      "longitude": -96.7872,
      "walkingDistance": "10m",
      "pic_url": "https://images.greenvillepizzaco.com/restaurant.jpg",
      "amenities": ["Vegetarian Options", "Bike Parking", "Alcohol Options"],
      "stationId": 1
    },
    {
      "poiId": 3,
      "name": "Uptown Theater",
      "type": "Theater",
      "latitude": 32.8482,
      "longitude": -96.7873,
      "walkingDistance": "6m",
      "pic_url": "https://images.uptowntheater.com/theater.jpg",
      "amenities": ["Vegetarian Options", "Bike Parking", "Alcohol Options"],
      "stationId": 3
    }
  ]
}
```
</details>

## **GET:** `public/line`

Fetches information about specific rail lines, including the Stations they connect to, based on the line color.

**Request Parameters**

 - Color (required): The color of the line you would like information for (as a string)

**Example Request**

`dallasbymetro.com/api/public/line?Color=BLUE`

<details>
<summary>Example Response</summary>

```json
{
  "line": {
    "color": "BLUE",
    "stations": [
      {
        "stationId": 1,
        "name": "Mockingbird Station"
      },
      {
        "stationId": 2,
        "name": "Cityplace/Uptown Station"
      }
    ]
  }
}
```
</details>

# Checklist 
- [x] /all
- [x] #175
- [x] #166
- [x] /line
- [x] #173

### LINKED ISSUE #175: /station

### PR DESCRIPTION
**Pull request by ZaneBartlett1**
closes #175 and closes #128

### CODE REVIEW

**Comment by ZaneBartlett1** [comment]
So two notes here -
- I did go ahead and add walk time as a filter, but it's optional, so you don't need to implement it. The only reason I did it is because I notice on the front end, it's not waiting for you to hit submit before it filters. I think it's a tad bit clunky to have it that way since none of the other filters have that. If/when that's updated I wasn't sure if it was easier to do on the front end or use the filter
- I did go ahead and just do it based on the number of connections rather than a rail time, like we discussed on a call. The RTFS data is very hard to parse that information out and due to how we got the station connections (not from the RTFS data) it's even harder.

**Comment by ZaneBartlett1** [comment]
I updated some stuff so that by default it sorts by level and then alphabetically within levels. This way the stations that are closest will show up near the top. The alphabetically sorting within levels is arbitrary.

File: src/main/java/com/dallasbymetro/backend/service/StationServiceImpl.java [hunk 1]
```diff
@@ -111,4 +110,93 @@ public List<PointOfInterestDTO> getPOIsById(List<Long> poiIds) throws DartExplor
...
+
+        return StationDTO.prepareStationDTO(station);
+    }
+
+    private List<Station> findStationsWithinConnection(Station currentStation, Integer stationConnections) {
+        Queue<Pair<Station, Integer>> queue = new LinkedList<>();
+        Map<Station, Integer> visitedMap = new HashMap<>();
+
+        // Create a comparator to sort the pairs by their level and station name
+        Comparator<Pair<Station, Integer>> pairComparator = Comparator
```
Comment (Nightwing52) [inline src/main/java/com/dallasbymetro/backend/service/StationServiceImpl.java:164]: Keep it this way because it is more general, but if you wanted to take advantage of the fact DART rail system is a tree (no loops), then each time you iterate you're guaranteed to go up a level.

**Review by Nightwing52 — CHANGES_REQUESTED** [review]
Filtering by walkDistance can be eliminated and other minor things. My main question is: what happens when no currentStation is provided to /stations? I believe it throws an error in the service layer, but it should do so by Bean validation.

File: src/main/java/com/dallasbymetro/backend/controller/ReadController.java [hunk 2]
```diff
@@ -76,10 +75,10 @@ ResponseEntity<List<PointOfInterestDTO>> getPOIs(@RequestParam("amenityIdList")
     ResponseEntity<List<PointOfInterestDTO>> getPOIsAtStation(@PathVariable Long stationId, @RequestParam("amenityIdList") String amenitiesString) throws DartExploreException {
         // probably move this first bit to a utility class later
         List<Long> amenityIdList;
-        try{
+        try {
             amenityIdList = (amenitiesString.isEmpty()) ? new ArrayList<>() :
                     Arrays.stream(amenitiesString.split(",")).map(Long::parseLong).collect(Collectors.toList());
-        }catch(Exception e){
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (ZaneBartlett1) [reply src/main/java/com/dallasbymetro/backend/controller/ReadController.java:81]: Yes, and I'm open to using something else. This is just default styling from JetBrains native to the IDE. You would just need to turn it on if that's what you're using, otherwise let me know if there's something else you'd want to set up

File: src/main/java/com/dallasbymetro/backend/utility/Pair.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.dallasbymetro.backend.utility;
+
+public class Pair<T1, T2> {
```
[the remark this answers is not in the document]
Comment (ZaneBartlett1) [reply src/main/java/com/dallasbymetro/backend/utility/Pair.java:3]: Ah, missed that. I'll keep it as is for now as you suggest, but thanks for letting me know

File: src/main/java/com/dallasbymetro/backend/controller/ReadController.java [hunk 4]
```diff
@@ -109,4 +108,31 @@ public ResponseEntity<List<String>> getAllTypes() {
         List<String> types = pointOfInterestService.getAllTypes();
         return new ResponseEntity<>(types, HttpStatus.OK);
     }
+
+    @GetMapping(value = "/stations")
+    public ResponseEntity<List<StationDTO>> getStationsByConnection(
+            @RequestParam(value = "currentStation", required = false) Long currentStation,
```
[the remark this answers is not in the document]
Comment (ZaneBartlett1) [reply src/main/java/com/dallasbymetro/backend/controller/ReadController.java:114]: Ah, good catch. Right now if you don't provide it, it'll send an error saying that a null station does not exist. I think technically that would be an acceptable error message, but if you're using the route both the current station and the maxStationConnections should be required.

[the remark this answers is not in the document]
Comment (ZaneBartlett1) [reply src/main/java/com/dallasbymetro/backend/controller/ReadController.java:114]: Updated both these fields to required

File: src/main/java/com/dallasbymetro/backend/service/StationServiceImpl.java [hunk 5]
```diff
@@ -111,4 +110,93 @@ public List<PointOfInterestDTO> getPOIsById(List<Long> poiIds) throws DartExplor
         }
         return poiList;
     }
+
+    @Override
+    public List<StationDTO> getStationsByConnection(Long currentStation, Integer stationConnections, List<Long> amenityIdList, Integer maxWalkTime) throws DartExploreException {
+        Optional<Station> stationOptional = stationRepository.findByStationIdIs(currentStation);
```
[the remark this answers is not in the document]
Comment (ZaneBartlett1) [reply src/main/java/com/dallasbymetro/backend/service/StationServiceImpl.java:116]: Don't think this really matters, but I changed it. It's going to be findByStationId though from JPL auto generated query. The original one was as well

File: src/main/java/com/dallasbymetro/backend/service/PointOfInterestService.java [hunk 6]
```diff
@@ -137,4 +137,21 @@ public List<PointOfInterestDTO> getPOIsByIds(List<Long> poiIdList) throws DartEx
...
+
+        for (PointOfInterest poi : pois) {
+            if (!poi.getAmenities().stream().map(Amenity::getAmenityId).collect(Collectors.toSet()).containsAll(amenityIds)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean doAllPOIsWithinWalkTime(List<PointOfInterest> pois, Integer maxWalkTime) {
```
[the remark this answers is not in the document]
Comment (ZaneBartlett1) [reply src/main/java/com/dallasbymetro/backend/service/PointOfInterestService.java:154]: I'd like to keep this as an option just because it does make since to include in the API. Since it's set to optional though you are free to disregard it. I'll address your issue below with how the API is filtering. 

Like I mention though I would like to have the front end updated so that it's filtered on button click rather than as the box is updated, since none of the other choices update on selection but rather on submit

File: src/main/java/com/dallasbymetro/backend/service/StationServiceImpl.java [hunk 7]
```diff
@@ -111,4 +110,93 @@ public List<PointOfInterestDTO> getPOIsById(List<Long> poiIds) throws DartExplor
...
+        List<Station> stationsWithinConnection = findStationsWithinConnection(station, stationConnections);
+
+        if (!amenityIdList.isEmpty()) {
+            // Filter stations by checking if the associated POIs have all the amenities specified.
+            stationsWithinConnection = stationsWithinConnection.stream()
+                    .filter(s -> PointOfInterestService.doAllPOIsHaveAmenities(s.getPointOfInterest(), amenityIdList))
+                    .collect(Collectors.toList());
+        }
+
+        if (maxWalkTime != null) {
```
[the remark this answers is not in the document]
Comment (ZaneBartlett1) [reply src/main/java/com/dallasbymetro/backend/service/StationServiceImpl.java:134]: Ah, fair enough. I've gone ahead and combined the filters.

File: src/main/java/com/dallasbymetro/backend/service/PointOfInterestService.java [hunk 8]
```diff
@@ -137,4 +137,21 @@ public List<PointOfInterestDTO> getPOIsByIds(List<Long> poiIdList) throws DartEx
     public List<String> getAllTypes() {
         return pointOfInterestRepository.getAllTypes();
     }
+
+    public static boolean doAllPOIsHaveAmenities(List<PointOfInterest> pois, List<Long> amenityIds) {
```
[the remark this answers is not in the document]
Comment (ZaneBartlett1) [reply src/main/java/com/dallasbymetro/backend/service/PointOfInterestService.java:141]: Great catch! I've updated the code so that now we should be filtering out at the POI level, and if a station has no POI it gets filtered out.

File: src/main/java/com/dallasbymetro/backend/controller/ReadController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ZaneBartlett1) [reply src/main/java/com/dallasbymetro/backend/controller/ReadController.java]: https://github.com/DartExplore/dallas-by-metro-backend/pull/209/commits/17961823e7c07ee32da2d38f3d6f35ddc30482d5

**Review by Nightwing52 — APPROVED** [review]
Approved.
